cpp/include/raft/neighbors/detail/knn_brute_force.cuh

```diff
@@ -548,4 +579,172 @@ void brute_force_search(
                                          norms.size() ? &norms : nullptr,
                                          query_norms ? query_norms->data_handle() : nullptr);
 }
+
+template <typename T, typename IdxT, typename BitmapT>
+void brute_force_search(
+  raft::resources const& res,
+  const raft::neighbors::brute_force::index<T>& idx,
+  raft::device_matrix_view<const T, IdxT, row_major> queries,
+  raft::core::bitmap_view<const BitmapT, IdxT> filter,
+  raft::device_matrix_view<IdxT, IdxT, row_major> neighbors,
+  raft::device_matrix_view<T, IdxT, row_major> distances,
+  std::optional<raft::device_vector_view<const T, IdxT>> query_norms = std::nullopt)
+{
+  auto metric = idx.metric();
+
+  RAFT_EXPECTS(neighbors.extent(1) == distances.extent(1), "Value of k must match for outputs");
+  RAFT_EXPECTS(idx.dataset().extent(1) == queries.extent(1),
+               "Number of columns in queries must match brute force index");
+  RAFT_EXPECTS(metric == raft::distance::DistanceType::InnerProduct ||
+                 metric == raft::distance::DistanceType::L2Expanded ||
+                 metric == raft::distance::DistanceType::L2SqrtExpanded ||
+                 metric == raft::distance::DistanceType::CosineExpanded,
+               "Only Euclidean, IP, and Cosine are supported!");
+
+  RAFT_EXPECTS(idx.has_norms() || !(metric == raft::distance::DistanceType::L2Expanded ||
+                                    metric == raft::distance::DistanceType::L2SqrtExpanded ||
+                                    metric == raft::distance::DistanceType::CosineExpanded),
+               "Index must has norms when using Euclidean, IP, and Cosine!");
+
+  IdxT n_queries = queries.extent(0);
+  IdxT n_dataset = idx.dataset().extent(0);
+  IdxT dim       = idx.dataset().extent(1);
+  IdxT k         = neighbors.extent(1);
+
+  auto stream = resource::get_cuda_stream(res);
+
+  // calc nnz
+  IdxT nnz_h = 0;
+  rmm::device_scalar<IdxT> nnz(0, stream);
+  auto nnz_view = make_device_scalar_view<IdxT>(nnz.data());
+  auto filter_view =
+    raft::make_device_vector_view<const BitmapT, IdxT>(filter.data(), filter.n_elements());
+
+  raft::detail::popc(res, filter_view, n_queries * n_dataset, nnz_view);
+  raft::copy(&nnz_h, nnz.data(), 1, stream);
+
+  resource::sync_stream(res, stream);
```

Comment:
Just a small note, and this applies for the whole PR- it'll be a lot easier to migrate this code to cuVS if you always use absolute namespaces (e..g `raft::resource::sync_stream`). It's usually a simple find/replace but sometimes it can get tricky when, for example RAFT has some overloads of existing intrinsics.

Reply:
Got it!